[JSC] Improve String#split for uglify-js-wtb

Constellation opened the pull request:
#### 8b7e7ed621bf5db6b4a46fb191f20d6d03a36589
<pre>
[JSC] Improve String#split for uglify-js-wtb
<a href="https://bugs.webkit.org/show_bug.cgi?id=251823">https://bugs.webkit.org/show_bug.cgi?id=251823</a>
rdar://105104781

Reviewed by Michael Saboff.

This patch improves JetStream2/uglify-js-wtb by 1% with String#split optimizations.

1. We should just use ArrayWithContigous array for result of String#split, and let&apos;s say `1` capacity at least.
   At that point, we will always return an array with at least one capacity. So, let&apos;s avoid structure transition,
   and butterfly reallocation.
2. We should search for BoyerMoore lookahead character patterns with Greedy one-character pattern. For example,
   /\r?\n/ is very frequently used. We should search for [\r\n] character to start matching in this case.
   However, if we have /\r?\n/ pattern,
       1. first character can be [\r\n]
       2. second character can be null, or [\n]
   So, this confuses fixed-sized BoyerMoore lookahead generation. So, when we encounter greedy one-character pattern, we cut
   the BM prefix length at this point. So, in the above case, we only consider about first character [\r\n] case.
   This is still beneficial since previously we give up completely when we encounter greedy fixed-sized one-character pattern.

                                   ToT                     Patched

    string-split-space      130.3261+-0.7798     ^    125.6028+-0.4402        ^ definitely 1.0376x faster
    string-split            206.8225+-1.1932     ^    128.5716+-0.7137        ^ definitely 1.6086x faster

* JSTests/microbenchmarks/string-split-space.js: Added.
(split):
* JSTests/microbenchmarks/string-split.js: Added.
(split):
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:

Canonical link: <a href="https://commits.webkit.org/259941@main">https://commits.webkit.org/259941@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2f17aa910b54337672cc90801da33dfa8b1de90

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106485 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15505 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39298 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115670 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16997 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6731 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98682 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112253 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95888 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94790 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27532 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96065 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8737 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28884 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95502 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6632 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9272 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/5456 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30638 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48430 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104243 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10818 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25828 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3698 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -505,7 +505,6 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoFuncSplitFast, (JSGlobalObject* globalObject
     // 11. Let A be ArrayCreate(0).
     // 12. Let lengthA be 0.
-    JSArray* result = constructEmptyArray(globalObject, nullptr);
     RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
We don't need this `RETURN_IF_EXCEPTION` anymore now that we don't `constructEmptyArray`.

Constellation replied:
Done

MenloDorian commented on Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
```diff
@@ -522,14 +521,15 @@ JSC_DEFINE_HOST_FUNCTION(regExpProtoFuncSplitFast, (JSGlobalObject* globalObject
         // b. If z is not null, return A.
         // c. Perform ! CreateDataProperty(A, "0", S).
         // d. Return A.
+        JSArray* result = constructEmptyArray(globalObject, nullptr);
```
Need a `RETURN_IF_EXCEPTION(scope, encodedJSValue());` after this.

Constellation replied:
Done

msaboff reviewed (approved):
r=me

Could you clarify the change comment that you are talking about Boyer-Moore lookahead characters and not a RegExp lookahead term?

Constellation committed:
[JSC] Improve String#split for uglify-js-wtb
https://bugs.webkit.org/show_bug.cgi?id=251823
rdar://105104781

Reviewed by Michael Saboff.

This patch improves JetStream2/uglify-js-wtb by 1% with String#split optimizations.

1. We should just use ArrayWithContigous array for result of String#split, and let's say `1` capacity at least.
   At that point, we will always return an array with at least one capacity. So, let's avoid structure transition,
   and butterfly reallocation.
2. We should search for BoyerMoore lookahead character patterns with Greedy one-character pattern. For example,
   /\r?\n/ is very frequently used. We should search for [\r\n] character to start matching in this case.
   However, if we have /\r?\n/ pattern,
       1. first character can be [\r\n]
       2. second character can be null, or [\n]
   So, this confuses fixed-sized BoyerMoore lookahead generation. So, when we encounter greedy one-character pattern, we cut
   the BM prefix length at this point. So, in the above case, we only consider about first character [\r\n] case.
   This is still beneficial since previously we give up completely when we encounter greedy fixed-sized one-character pattern.

                                   ToT                     Patched

    string-split-space      130.3261+-0.7798     ^    125.6028+-0.4402        ^ definitely 1.0376x faster
    string-split            206.8225+-1.1932     ^    128.5716+-0.7137        ^ definitely 1.6086x faster

* JSTests/microbenchmarks/string-split-space.js: Added.
(split):
* JSTests/microbenchmarks/string-split.js: Added.
(split):
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:

Canonical link: https://commits.webkit.org/259941@main
